File: concepts/08_async_and_concurrency/01_async_await_and_result_propagation/lua/test_02_task_ownership_and_failure_collection.lua
-- Common question: who owns a spawned unit and observes its failure and cleanup?
-- Inputs: a suspended coroutine, an unobserved error, coroutine.close, and coroutine.wrap.
-- Observations: caller-owned resume, explicit failure collection, deterministic close, and wrap propagation.
-- polyglot-family: async_and_concurrency
-- polyglot-concept: async_await_and_result_propagation
-- polyglot-related: languages/lua/language/08_coroutines_and_gc/
-- polyglot-related+: test_060_coroutine_errors_wrap_and_explicit_close.lua

local t = require("support.assertions")
local closed = 0
local resource = setmetatable({}, {__close = function() closed = closed + 1 end})
local worker = coroutine.create(function()
    local handle <close> = resource
    t.same(handle, resource)
    coroutine.yield("owned")
    error("worker failed")
end)

t.pack_equal(table.pack(coroutine.resume(worker)), {n = 2, true, "owned"})
local ok, error_value = coroutine.resume(worker)
t.falsey(ok)
t.matches(error_value, "worker failed")
t.equal(closed, 0)
local close_ok, close_error = coroutine.close(worker)
t.falsey(close_ok)
t.matches(close_error, "worker failed")
t.equal(closed, 1)

local wrapped = coroutine.wrap(function() error("wrapped failure") end)
t.raises(wrapped, "wrapped failure")

t.done()

File: concepts/08_async_and_concurrency/01_async_await_and_result_propagation/ruby/test_02_task_ownership_and_failure_collection.rb
# 共同问题：谁拥有启动的工作单元，并负责观察失败和清理。
# 输入：Thread、ensure、显式 join/value 和未启用 abort_on_exception；观察：caller 收集结果及 worker 清理。
# polyglot-family: async_and_concurrency
# polyglot-concept: async_await_and_result_propagation
# polyglot-related: languages/ruby/standard_library/11_concurrency_models/
# polyglot-related+: test_081_thread_results_shared_state_and_failures.rb

require "assertions"

A = PolyglotAssertions

trace = Queue.new
worker = Thread.new do
  begin
    trace << :started
    raise "owned failure"
  ensure
    trace << :cleaned
  end
end
worker.report_on_exception = false
A.raises(RuntimeError, "owned failure") { worker.join }
A.equal(:started, trace.pop)
A.equal(:cleaned, trace.pop)
A.falsey(worker.alive?)
A.raises(RuntimeError, "owned failure") { worker.value }
A.falsey(Thread.abort_on_exception)

A.done
